src/transformStateWithClones.js

```diff
@@ -7,7 +7,38 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const result = [];
+  let current = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties': {
+        current = {
+          ...current, ...action.extraData,
+        };
+        break;
+      }
+
+      case 'removeProperties': {
+        for (const key of action.keysToRemove) {
+          delete current[key];
+        }
+        break;
+      }
+
+      case 'clear': {
+        current = {};
+        break;
+      }
+
+      default: {
+        break;
+      }
+    }
+    result.push({ ...current });
```

Comment:
The cloning of the 'current' object is unnecessary in this line as it is already cloned inside the 'switch' statement.